### PR TITLE
Feature/nk initial commit

### PR DESCRIPTION
**Pull request by krakover**

### CODE REVIEW

File: cloudstorage/errors.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        if job_ref:
+            message = 'Error processing %r: %s' % (job_ref, error.get('message'))
+        else:
+            message = error.get('message')
+
+
+        new_errors = [err for err in error_ls if err != error]
+        if new_errors:
+            message += '\nFailure details:\n'
+            message += '\n'.join(textwrap.fill(': '.join(filter(None, [err.get('location', None), err.get('message', '')])), initial_indent=' - ', subsequent_indent='   ') for err in new_errors)
```
Comment (ekampf) [inline cloudstorage/errors.py]: Wow. what does this line do? :)

File: cloudstorage/errors.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            return GoogleCloudStorageError('Error reported by server with missing error fields. ' 'Server returned: %s' % (str(server_error),))
+
+        if reason == 'authError':
+            return GoogleCloudStorageAuthorizationError(message)
+
+        # We map the less interesting errors to GoogleCloudStorageServiceError.
+        return GoogleCloudStorageServiceError(message, error, error_ls, job_ref=job_ref)
+
+
+class GoogleCloudStorageAuthorizationError(GoogleCloudStorageError):
```
Comment (ekampf) [inline cloudstorage/errors.py]: 401 is Authentication error ("I don't know who you are")

Authorization error is when server returns 403 Forbidden ("I know who you are but you don't have permission to do this")

Comment (ekampf) [inline cloudstorage/errors.py]: I saw on client.py that you are using this correctly when code is 403.
So just the doc that says 401 is wrong?

Comment (ekampf) [inline cloudstorage/errors.py]: Shouldn't there be a separate 401 exception class (and relevant tests in client.py) ?

File: cloudstorage/client.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            metadata_uri = 'http://metadata.google.internal'
+            http = httplib2.Http()
+            http.request(metadata_uri, method='GET')
+            return True
+        except httplib2.ServerNotFoundError:
+            return False
+
+    @property
+    def api_client(self):
+        http = self.get_http_for_request()
```
Comment (ekampf) [inline cloudstorage/client.py]: Consider using the same method I use in PyBigQuery overriding the Http request and model objects.
This way you can process all results for errors in one place and not have to catch and process exceptions on each method separately (which is bug prone)

```
@property
    def api_client(self):
        bigquery_model = BigQueryModel(trace=self.trace)
        bigquery_http = BigQueryHttp.factory(bigquery_model)

        http = self.get_http_for_request()
        return build("bigquery", "v2", http=http, model=bigquery_model, requestBuilder=bigquery_http)
```

File: cloudstorage/client.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            logging.info("Using Standard jwt authentication")
+            self._credentials = self._credentials
+            return credentials
+        elif self.is_in_appengine():  # App engine
+            scope = 'https://www.googleapis.com/auth/bigquery'
+            credentials = AppAssertionCredentials(scope=scope)
+            logging.info("Using Standard appengine authentication")
+            self._credentials = self._credentials
+            return credentials
+        elif self.oauth_credentails_file:  # Local oauth token
```
Comment (ekampf) [inline cloudstorage/client.py]: Shouldn't local be last in the options? (after GCE)

**Comment by ekampf** [comment]
Consider adding .pylintrc + Makefile?

File: cloudstorage/client.py (hunk 4 above)
Comment (krakover) [inline cloudstorage/client.py]: No, because you might want to override it, even if you are in a gce machine

File: cloudstorage/errors.py (hunk 2 above)
Comment (krakover) [inline cloudstorage/errors.py]: This case is relevant for jwt authentication? (in appengine/oauth2/gce authentication you are not a person, but an application. You might not be able to do specific actions, but you are either known or not allowed in this API).

**Pull request merged**